govcd/nsxt_edgegateway.go

```diff
@@ -359,3 +374,341 @@ func filterOnlyNsxtEdges(allEdges []*NsxtEdgeGateway) []*NsxtEdgeGateway {
+// this subnet only.
+// * `refresh` defines if Edge Gateway structure should be retrieved with latest data before
+// performing IP lookup operation
+//
+// Input and return arguments are using Go's native 'netip' package for IP addressing. This ensures
+// correct support for IPv4 and IPv6 IPs.
+// `netip.ParseAddr`, `netip.ParsePrefix`, `netip.Addr.String` functions can be used for conversion
+// from/to strings
+//
+// This function performs below listed steps:
+// 1. Retrieves a complete list of IPs in Edge Gateway uplinks (returns error if none are found)
+// 2. if 'optionalSubnet' was specified - filter IP addresses to only fall into that subnet
+// 3. Retrieves all used IP addresses in Edge Gateway using dedicated API endpoint
+// 4. Subtracts used IP addresses from available list of IPs in uplink (optionally filtered by optionalSubnet in step 2)
+// 5. Checks if 'requiredIpCount' criteria is met, returns error otherwise
+// 6. Returns required amount of unused IPs (as defined in 'requiredIpCount')
+//
+// Notes:
+// * This function uses Go's builtin `netip` package to avoid any string processing of IPs and
+// supports IPv4 and IPv6 addressing.
+// * If an unused IP is not found it will return 'netip.Addr{}' (not using *netip.Addr{} to match
+// library semantics) and an error
+// * It will return an error if any of uplink IP ranges End IP address is lower than Start IP
+// address
+func (egw *NsxtEdgeGateway) GetUnusedExternalIPAddresses(requiredIpCount int, optionalSubnet netip.Prefix, refresh bool) ([]netip.Addr, error) {
+	if refresh {
+		err := egw.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing Edge Gateway: %s", err)
+		}
+	}
+	usedIpAddresses, err := egw.GetUsedIpAddresses(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error getting used IP addresses for Edge Gateway: %s", err)
+	}
+
+	return getUnusedExternalIPAddress(egw.EdgeGateway.EdgeGatewayUplinks, usedIpAddresses, requiredIpCount, optionalSubnet)
+}
+
+// GetAllUnusedExternalIPAddresses will retrieve all unassigned IP addresses for Edge Gateway It is
+// similar to GetUnusedExternalIPAddresses but returns all unused IPs instead of a specific amount
+func (egw *NsxtEdgeGateway) GetAllUnusedExternalIPAddresses(refresh bool) ([]netip.Addr, error) {
+	if refresh {
+		err := egw.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing Edge Gateway: %s", err)
+		}
+	}
+	usedIpAddresses, err := egw.GetUsedIpAddresses(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error getting used IP addresses for Edge Gateway: %s", err)
+	}
+
+	return getAllUnusedExternalIPAddresses(egw.EdgeGateway.EdgeGatewayUplinks, usedIpAddresses, netip.Prefix{})
+}
+
+// GetAllocatedIpCount traverses all subnets in Edge Gateway and returns a sum of allocated ip count
+// for each subnet in each uplink
+func (egw *NsxtEdgeGateway) GetAllocatedIpCount(refresh bool) (*int, error) {
```

Comment:
Well spotted. I think during the initial design my main thought was having the option to distinguish between `0` and `nil`, but it is pretty much impossible to have `0` allocated IPs for an Edge Gateway (because it must always have at least a primary IP address).

I have changed this now.